src/symbol/collision_index.js

```diff
@@ -143,99 +124,119 @@ class CollisionIndex {
-                    if (!showCollisionCircles) {
-                        return {
-                            circles: [],
-                            offscreen: false
-                        };
-                    } else {
-                        // Don't early exit if we're showing the debug circles because we still want to calculate
-                        // which circles are in use
-                        collisionDetected = true;
+
+            for (const seg of segments) {
+                // interpolate positions for collision circles. Add a small padding to both ends of the segment
+                interpolator.reset(seg, radius * 0.5);
+
+                let numCircles = 0;
+
+                if (interpolator.length <= radius) {
+                    numCircles = 1;
+                } else {
+                    numCircles = Math.ceil(interpolator.paddedLength / circleDist) + 1;
+                }
+
+                for (let i = 0; i < numCircles; i++) {
+                    const t = i / Math.max(numCircles - 1, 1);
+                    const circlePosition = interpolator.lerp(t);
+
+                    if (!circlePosition)
+                        continue;   // To silence flow
+
+                    // add viewport padding to the position and perform initial collision check
+                    const centerX = circlePosition.x + viewportPadding;
+                    const centerY = circlePosition.y + viewportPadding;
+
+                    placedCollisionCircles.push(centerX, centerY, radius, 0);
+
+                    const x1 = centerX - radius;
+                    const y1 = centerY - radius;
+                    const x2 = centerX + radius;
+                    const y2 = centerY + radius;
+
+                    entirelyOffscreen = entirelyOffscreen && this.isOffscreen(x1, y1, x2, y2);
+                    inGrid = inGrid || this.isInsideGrid(x1, y1, x2, y2);
+
+                    if (!allowOverlap) {
+                        if (this.grid.hitTestCircle(centerX, centerY, radius, collisionGroupPredicate)) {
+                            // Don't early exit if we're showing the debug circles because we still want to calculate
+                            // which circles are in use
+                            collisionDetected = true;
+                            if (!showCollisionCircles) {
+                                return {
+                                    circles: [],
+                                    offscreen: false,
+                                    collisionDetected
+                                };
+                            }
+                        }
                     }
                 }
             }
         }
 
         return {
-            circles: (collisionDetected || !inGrid) ? [] : placedCollisionCircles,
-            offscreen: entirelyOffscreen
+            circles: ((!showCollisionCircles && collisionDetected) || !inGrid) ? [] : placedCollisionCircles,
```

Comment:
is there a need to construct an empty `[]` array?

Reply:
I think it could be replaced with null